File: Classes/AMGAboutViewController.m
//
//  AMGAboutViewController.m
//
//  Created by Vincent Tourraine on 11/03/2019.
//  Copyright © 2019-2021 Studio AMANgA. All rights reserved.
//

#import "AMGAboutViewController.h"
#import "AMGAboutAction.h"
#import "AMGAboutHeaderView.h"
#import "AMGAboutFooterView.h"

#import <SVProgressHUD.h>
#import <AMGAppButton/AMGApp.h>
#import <AMGAppButton/AMGAppButton.h>
#import <VTAppButton.h>
#import <VTAcknowledgementsViewController.h>
#import <VTLocalization.h>


@implementation AMGAboutViewController

#pragma mark - View life cycle

- (void)commonInit {
    self.title = NSLocalizedString(@"About this App", nil);

    AMGSettingsDataRow *reviewRow = [[AMGSettingsDataRow alloc] initWithTitle:NSLocalizedString(@"Review on the App Store", nil) imageName:@"IconStar" systemImageName:@"star" action:^(id sender) {
        [self reviewOnAppStore:sender];
    }];
    AMGSettingsDataRow *shareRow = [[AMGSettingsDataRow alloc] initWithTitle:NSLocalizedString(@"Share the App", nil) imageName:@"IconChat" systemImageName:@"square.on.square" action:^(id sender) {
        [self shareApp:sender];
    }];
    AMGSettingsDataRow *twitterRow = [[AMGSettingsDataRow alloc] initWithTitle:NSLocalizedString(@"Follow on Twitter", nil) imageName:@"IconTwitter" systemImageName:@"text.bubble" action:^(id sender) {
        [self openTwitterAccount:sender];
    }];
    self.rows = @[reviewRow, shareRow, twitterRow];

    AMGSettingsAction *ackRow = [[AMGSettingsAction alloc] initWithTitle:[VTLocalization localizedTitle] action:^(AMGAboutViewController *viewController) {
        [viewController presentLicensesViewController:nil];
    }];
    self.footerActions = @[ackRow];
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    [self commonInit];

    return self;
}

- (instancetype)init {
    if (@available(iOS 13.0, *)) {
        self = [super initWithStyle:UITableViewStyleInsetGrouped];
    }
    else {
        self = [super initWithStyle:UITableViewStyleGrouped];
    }

    [self commonInit];

    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    [self configureTableHeaderAndFooter];

    [self.tableView registerClass:UITableViewCell.class forCellReuseIdentifier:@"Cell"];
    self.tableView.cellLayoutMarginsFollowReadableWidth = YES;

    if (@available(iOS 11.0, *)) {
        self.navigationController.navigationBar.prefersLargeTitles = YES;
    }
}

#pragma mark - Configuration

- (void)configureTableHeaderAndFooter {
    if (self.largeIconName != nil) {
        self.tableView.tableHeaderView = [[AMGAboutHeaderView alloc] initWithIconImageNamed:self.largeIconName];
    }

#if TARGET_OS_MACCATALYST
    self.tableView.tableFooterView = [AMGAboutFooterView creditsLabelForViewController:self];
#else
    NSArray <AMGApp *> *allApps = @[AMGApp.appGamesKeeper, AMGApp.appComicBookDay, AMGApp.appContacts, AMGApp.app1List, AMGApp.appWizBox, AMGApp.appMemorii, AMGApp.appMegaMoji, AMGApp.appD0TSEchoplex, AMGApp.appNanoNotes];
    NSArray <AMGApp *> *otherApps = [allApps filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"%K != %@", NSStringFromSelector(@selector(identifier)), self.appIdentifier]];

    self.tableView.tableFooterView = [[AMGAboutFooterView alloc] initForViewController:self withApps:otherApps actions:self.footerActions];
#endif
}

#pragma mark - Actions

- (IBAction)reviewOnAppStore:(id)sender {
    if (self.appIdentifier == nil) {
        return;
    }

#if TARGET_OS_MACCATALYST
    NSString *URLString = [NSString stringWithFormat:@"macappstore://apps.apple.com/app/id%@?action=write-review", self.appIdentifier];
#else
    NSString *URLString = [NSString stringWithFormat:@"itms-apps://itunes.apple.com/app/id%@?action=write-review", self.appIdentifier];
#endif

    NSURL *URL = [NSURL URLWithString:URLString];
    [[UIApplication sharedApplication] openURL:URL options:@{} completionHandler:nil];
}

- (void)shareApp:(id)sender {
    if (self.localizedAppName == nil || self.shareAppURL == nil) {
        return;
    }

    NSArray *items = @[self.localizedAppName, self.shareAppURL];
    UIActivityViewController *activityViewController = [[UIActivityViewController alloc] initWithActivityItems:items applicationActivities:nil];

    activityViewController.popoverPresentationController.sourceView = self.tableView;
    if ([sender isKindOfClass:NSIndexPath.class]) {
        activityViewController.popoverPresentationController.sourceRect = [self.tableView rectForRowAtIndexPath:sender];
    }

    [self presentViewController:activityViewController animated:YES completion:nil];
}

- (IBAction)openTwitterAccount:(id)sender {
    UIApplication *application = [UIApplication sharedApplication];
    NSURL *twitterURL = [NSURL URLWithString:@"https://www.twitter.com/StudioAMANgA"];
    [application openURL:twitterURL options:@{} completionHandler:nil];
}


+ (void)presentContactSupportViewControllerFrom:(UIViewController <MFMailComposeViewControllerDelegate> *)presentingViewController {
    [self presentContactSupportViewControllerFrom:presentingViewController delegate:presentingViewController];
}

+ (void)presentContactSupportViewControllerFrom:(UIViewController *)presentingViewController delegate:(id<MFMailComposeViewControllerDelegate>)delegate {
    if (MFMailComposeViewController.canSendMail == NO) {
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Cannot Send Email", nil) message:NSLocalizedString(@"Please configure an email account first, or contact me directly at studioamanga@gmail.com", nil) preferredStyle:UIAlertControllerStyleAlert];
        [alertController addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK", nil) style:UIAlertActionStyleCancel handler:nil]];
        [presentingViewController presentViewController:alertController animated:YES completion:nil];

        return;
    }

    NSDictionary *bundleInfo = NSBundle.mainBundle.infoDictionary;
    MFMailComposeViewController *viewController = [[MFMailComposeViewController alloc] init];
    viewController.mailComposeDelegate = delegate;
    [viewController setToRecipients:@[@"studioamanga@gmail.com"]];
    NSString *bundleDisplayName = bundleInfo[@"CFBundleDisplayName"];
    NSString *bundleShortVersion = bundleInfo[@"CFBundleShortVersionString"];
    [viewController setSubject:[NSString stringWithFormat:@"%@ (%@)", bundleDisplayName, bundleShortVersion]];

    [presentingViewController presentViewController:viewController animated:YES completion:nil];
}

- (IBAction)showApplication:(id)sender {
    if ([sender isKindOfClass:VTAppButton.class] == NO) {
        return;
    }

    [SVProgressHUD show];

    NSString *appIdentifier = ((VTAppButton *)sender).appIdentifier;
    SKStoreProductViewController *storeProductViewController = [[SKStoreProductViewController alloc] init];
    storeProductViewController.delegate = self;
    [storeProductViewController loadProductWithParameters:@{SKStoreProductParameterITunesItemIdentifier: appIdentifier} completionBlock:^(BOOL result, NSError *error) {
        [SVProgressHUD dismiss];

        if (error) {
            NSLog(@"%@", error);
        } else {
            [self presentViewController:storeProductViewController animated:YES completion:nil];
        }
    }];
}

- (IBAction)presentLicensesViewController:(id)sender {
    UIViewController *viewController = nil;
    if (self.acknowledgementsFileName != nil ) {
        viewController = [[VTAcknowledgementsViewController alloc] initWithFileNamed:self.acknowledgementsFileName];
    }
    else {
        viewController = [VTAcknowledgementsViewController acknowledgementsViewController];
    }

    [self.navigationController pushViewController:viewController animated:YES];
}

- (IBAction)performFooterAction:(UIButton *)sender {
    AMGSettingsAction *action = self.footerActions[sender.tag];
    action.action(self);
}

- (IBAction)dismissViewController:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - Store product view delegate

- (void)productViewControllerDidFinish:(SKStoreProductViewController *)viewController {
    [viewController dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - Mail compose view delegate

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error {
    [controller dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.rows.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
    AMGSettingsDataRow *row = [self rowAtIndexPath:indexPath];
    cell.textLabel.text = row.title;

    if (@available(iOS 13.0, *)) {
        if (row.systemImageName) {
            cell.imageView.image = [UIImage systemImageNamed:row.systemImageName];
        }
        else {
            cell.imageView.image = [UIImage imageNamed:row.imageName];
        }
    }
    else {
        cell.imageView.image = [UIImage imageNamed:row.imageName];
    }

    return cell;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    AMGSettingsDataRow *row = [self rowAtIndexPath:indexPath];
    row.action(indexPath);

    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - Helpers

- (nullable AMGSettingsDataRow *)rowAtIndexPath:(NSIndexPath *)indexPath {
    return self.rows[indexPath.row];
}

@end

File: Sources/AMGAboutViewController/AMGAboutViewController.swift
//
//  AMGAboutViewController.swift
//
//  Created by Vincent Tourraine on 11/03/2019.
//  Copyright © 2019-2022 Studio AMANgA. All rights reserved.
//

#if !os(macOS)

import UIKit
import MessageUI
import StoreKit
import AcknowList
import VTAppButton
import AMGAppButton
import TrackupVersionHistory

public class AMGAboutViewController: UITableViewController {
    
    @objc public var localizedAppName: String?
    @objc public var largeIconName: String?
    @objc public var acknowledgementsFileName: String?
    @objc public var shareAppURL: URL?

    public var appIdentifier: Int?
    @objc public func setAppIdentifier(_ appIdentifier: NSNumber?) {
        self.appIdentifier = appIdentifier?.intValue
    }

    var rows: [AMGSettingsDataRow]?
    var footerActions: [AMGSettingsAction]?
    
    struct K {
        static let cellIdentifier = "Cell"
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        commonInit()
    }
    
    public init() {
        super.init(style: .insetGrouped)
        commonInit()
    }
    
    func commonInit() {
        title = NSLocalizedString("About this App", comment: "")
        
        let reviewRow = AMGSettingsDataRow(title: NSLocalizedString("Review on the App Store", comment: ""), systemImageName: "star") {
            self.reviewOnAppStore($0)
        }
        
        let shareRow = AMGSettingsDataRow(title: NSLocalizedString("Share the App", comment: ""), systemImageName: "square.on.square") {
            self.shareApp($0)
        }
        
        let twitterRow = AMGSettingsDataRow(title: NSLocalizedString("Follow on Twitter", comment: ""), systemImageName: "text.bubble") {
            self.openTwitterAccount($0)
        }
        
        let versionHistoryRow = AMGSettingsDataRow(title: NSLocalizedString("Version History", comment: ""), systemImageName: "list.bullet.rectangle") {
            self.presentVersionHistory($0)
        }

        rows = [reviewRow, shareRow, twitterRow, versionHistoryRow]
        footerActions = []
    }
    
    // MARK: - View life cycle
    
    public override func viewDidLoad() {
        super.viewDidLoad()

        if acknowledgementsFileName != nil {
            let ackAction = AMGSettingsAction(title: AcknowLocalization.localizedTitle()) { viewController in
                viewController.presentLicensesViewController(nil)
            }

            footerActions?.insert(ackAction, at: 0)
        }

        configureTableHeaderAndFooter()
        
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: K.cellIdentifier)
        tableView.cellLayoutMarginsFollowReadableWidth = true
        
        navigationController?.navigationBar.prefersLargeTitles = true
    }
    
    // MARK: - Configuration
    
    func configureTableHeaderAndFooter() {
        if let largeIconName = largeIconName {
            tableView.tableHeaderView = AMGAboutHeaderView(iconImageNamed: largeIconName)
        }
        
#if targetEnvironment(macCatalyst)
        tableView.tableFooterView = AMGAboutFooterView.creditsLabel(for: self)
#else
        let allApps: [AMGApp] = [.gamesKeeper, .comicBookDay, .contacts, .oneList, .wizBox, .memorii, .megaMoji, .d0tsEchoplex, .nanoNotes]
        let otherApps = allApps.filter { $0.identifier != appIdentifier }
        tableView.tableFooterView = AMGAboutFooterView(for: self, with: otherApps, actions: footerActions ?? [])
#endif
    }
    
    // MARK: - Actions
    
    func reviewOnAppStore(_ sender: Any) {
        guard let appIdentifier = appIdentifier else {
            return
        }
        
#if targetEnvironment(macCatalyst)
        let URLString = "macappstore://apps.apple.com/app/id\(appIdentifier)?action=write-review"
#else
        let URLString = "itms-apps://itunes.apple.com/app/id\(appIdentifier)?action=write-review"
#endif
        
        let url = URL(string: URLString)!
        UIApplication.shared.open(url, options: [:], completionHandler: nil)
    }
    
    func shareApp(_ sender: Any) {
        guard let localizedAppName = localizedAppName,
              let shareAppURL = shareAppURL else {
            return
        }
        
        let items = [localizedAppName, shareAppURL] as [Any]
        let activityViewController = UIActivityViewController(activityItems: items, applicationActivities: nil)
        activityViewController.popoverPresentationController?.sourceView = tableView
        
        if let sender = sender as? IndexPath {
            activityViewController.popoverPresentationController?.sourceRect = tableView.rectForRow(at: sender)
        }
        
        present(activityViewController, animated: true)
    }
    
    func openTwitterAccount(_ sender: Any) {
        let twitterURL = URL(string: "https://www.twitter.com/StudioAMANgA")!
        UIApplication.shared.open(twitterURL, options: [:], completionHandler: nil)
    }
    
    func presentVersionHistory(_ sender: Any) {
        let viewController = VersionHistoryViewController()
        navigationController?.pushViewController(viewController, animated: true)
    }

    // MARK: -
    
    @objc public static func presentContactSupportViewController(from presentingViewController: UIViewController & MFMailComposeViewControllerDelegate) {
        presentContactSupportViewController(from: presentingViewController, delegate: presentingViewController)
    }
    
    @objc public static func presentContactSupportViewController(from presentingViewController: UIViewController, delegate: MFMailComposeViewControllerDelegate) {
        guard MFMailComposeViewController.canSendMail() else {
            let alertController = UIAlertController(title: NSLocalizedString("Cannot Send Email", comment: ""), message: NSLocalizedString("Please configure an email account first, or contact me directly at studioamanga@gmail.com", comment: ""), preferredStyle: .alert)
            alertController.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""), style: .cancel))
            presentingViewController.present(alertController, animated: true)
            
            return
        }
        
        guard let bundleInfo = Bundle.main.infoDictionary,
              let bundleDisplayName = bundleInfo["CFBundleDisplayName"],
              let bundleShortVersion = bundleInfo["CFBundleShortVersionString"] else {
            return
        }
        
        let viewController = MFMailComposeViewController()
        viewController.mailComposeDelegate = delegate
        viewController.setToRecipients(["studioamanga@gmail.com"])
        viewController.setSubject("\(bundleDisplayName) (\(bundleShortVersion))")
        presentingViewController.present(viewController, animated: true)
    }
    
    @objc func showApplication(_ sender: Any) {
        guard let sender = sender as? AMGAppButton,
              let senderAppIdentifier = sender.appIdentifier else {
            return
        }
        
        sender.isLoading = true
        
        let appIdentifier = NSNumber(integerLiteral: senderAppIdentifier)
        let storeProductViewController = SKStoreProductViewController()
        storeProductViewController.delegate = self
        storeProductViewController.loadProduct(withParameters: [SKStoreProductParameterITunesItemIdentifier: appIdentifier]) { result, error in
            sender.isLoading = false

            if let error = error {
                print("\(error.localizedDescription)")
            }
            else {
                self.present(storeProductViewController, animated: true)
            }
        }
    }
    
    func presentLicensesViewController(_ sender: Any?) {
        let viewController: AcknowListViewController
        if let acknowledgementsFileName = acknowledgementsFileName {
            viewController = AcknowListViewController(fileNamed: acknowledgementsFileName)
        }
        else {
            viewController = AcknowListViewController()
        }

        navigationController?.pushViewController(viewController, animated: true)
    }
    
    @objc func performFooterAction(_ sender: UIButton) {
        if let action = footerActions?[sender.tag] {
            action.action(self)
        }
    }
    
    func dismissViewController(_ sender: Any) {
        dismiss(animated: true, completion: nil)
    }
    
    // MARK: - Table view data source
    
    public override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return rows?.count ?? 0
    }
    
    public override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: K.cellIdentifier, for: indexPath)
        if let row = row(at: indexPath) {
            cell.textLabel?.text = row.title
            cell.imageView?.image = UIImage(systemName: row.systemImageName)
        }
        
        return cell
    }
    
    // MARK: - Table view delegate
    
    func row(at indexPath: IndexPath) -> AMGSettingsDataRow? {
        return rows?[indexPath.row]
    }
    
    public override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if let row = row(at: indexPath) {
            row.action(indexPath)
        }
        
        tableView.deselectRow(at: indexPath, animated: true)
    }
}

extension AMGAboutViewController: MFMailComposeViewControllerDelegate {
    public func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        controller.dismiss(animated: true, completion: nil)
    }
}

extension AMGAboutViewController: SKStoreProductViewControllerDelegate {
    public func productViewControllerDidFinish(_ viewController: SKStoreProductViewController) {
        viewController.dismiss(animated: true, completion: nil)
    }
}

// Objective-C helpers
extension AMGAboutViewController {
    @objc public func addFooterAction(title: String, action: @escaping (AMGAboutViewController)->Void) {
        let action = AMGSettingsAction(title: title, action: action)
        footerActions?.append(action)
    }
}

#endif
